src/main/java/uk/nhs/hee/tis/trainee/credentials/filter/SignedDataFilter.java

```diff
@@ -0,0 +1,125 @@
+ * Copyright 2023 Crown Copyright (Health Education England)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+ * associated documentation files (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge, publish, distribute,
+ * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.tis.trainee.credentials.filter;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.time.Instant;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.codec.digest.HmacAlgorithms;
+import org.apache.commons.codec.digest.HmacUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+/**
+ * A request filter that verifies the signature of signed data.
+ */
+@Slf4j
+@Component
+public class SignedDataFilter extends OncePerRequestFilter {
+
+  private static final String SIGNATURE_FIELD = "signature";
+  private static final String HMAC_FIELD = "hmac";
+
+  private final ObjectMapper mapper;
+  private final String signatureSecretKey;
+
+  /**
+   * Create a request filter that verifies the signature of signed data.
+   *
+   * @param mapper             An {@link ObjectMapper} to use when reading the request payload.
+   * @param signatureSecretKey The secret key used to sign and verify the signature.
+   */
+  SignedDataFilter(ObjectMapper mapper,
+      @Value("${application.signature.secret-key}") String signatureSecretKey) {
+    this.mapper = mapper;
+    this.signatureSecretKey = signatureSecretKey;
+  }
+
+  @Override
+  protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
+      FilterChain filterChain) throws ServletException, IOException {
+    request = new CachedBodyRequestWrapper(request);
+
+    try {
```

Comment:
I wonder if a 403 is appropriate in all cases? I'm thinking a no-data or not-json request could possibly trigger a 400 instead. It probably doesn't matter much, but I thought worth flagging-up for consideration.

Reply:
I want to look at returning a more reasonable error message, but I think it has to be implemented using interceptors so kept it simple to start with.

Reply:
Ok, fair enough 👍